01-data-structures/07-kevin-bacon/kb_finder.rb

```diff
@@ -7,32 +7,48 @@ def find_kevin_bacon(node)
 
   queue = [node]
   path = {}
-  current_node = node
-  path[current_node.name] = {:distance => 0, 
-                             :previous_node => nil, 
-                             :encountered_movie => nil}
+  current_actor = node
+  path[current_actor.name] = {:distance => 0, 
+                              :previous_node => nil, 
+                              :encountered_movie => nil}
 
-  while (!queue.empty? and path[current_node.name][:distance] < 5)
-    current_node = queue.shift
-    current_node.film_actor_hash.each do |movie_title, actors|
-      for actor in actors
-        if actor.name == "Kevin Bacon"
-          return trace_path(movie_title, path, current_node)
-        elsif !path.has_key?(actor.name)
-          queue.push(actor)
-          path[actor.name] = {:distance => path[current_node.name][:distance] + 1, 
-                              :previous_node => path[current_node.name], 
-                              :encountered_movie => movie_title}
-        end
-      end
+  while (!queue.empty? and path[current_actor.name][:distance] < 5)
+    current_actor = queue.shift
+    knows_kevin_bacon = current_actor_knows_kevin_bacon(current_actor, queue, path)
+    if knows_kevin_bacon[0]
+      return knows_kevin_bacon[1]
     end
   end
   nil
 end
 
-def trace_path(movie_title, path, current_node)
-  answer = [movie_title]
-  prior_path_step = path[current_node.name]
+def current_actor_knows_kevin_bacon(current_actor, queue, path)
+  current_actor.film_actor_hash.each do |current_movie, actors|
+    knows_kevin_bacon = current_actor_worked_with_kevin_bacon_in_current_movie(current_movie, current_actor, actors, queue, path)
```

Comment:
Yes, you're right. I refactored a bit to _not_ return an array now. Don't know what I was thinking! Must have been coding for way too long.